types/prop-types/index.d.ts

```diff
@@ -1,33 +1,61 @@
 // Type definitions for prop-types 15.5
 // Project: https://github.com/reactjs/prop-types
 // Definitions by: DovydasNavickas <https://github.com/DovydasNavickas>
+//                 Ferdy Budhidharma <https://github.com/ferdaber>
 // Definitions: https://github.com/DefinitelyTyped/DefinitelyTyped
-// TypeScript Version: 2.2
+// TypeScript Version: 2.8
 
-export type Validator<T> = (object: T, key: string, componentName: string, ...rest: any[]) => Error | null;
+import { ReactElement, ReactNode } from 'react';
+
+declare const resolvedType: unique symbol;
+
+export interface Validator<T, R = false> {
+    (object: T, key: string, componentName: string, ...rest: any[]): Error | null;
+    [resolvedType]?: R extends true ? T : T | null | undefined;
```

Comment:
This is how I wrote it in my example, but it's a bit messy.  It's cleaner to just have a nullable validator and a separate strict validator type (rather than having a boolean type parameter), I think.

Reply:
Great! I did feel like it was clunky to implement it like this, it's kind of how the prop-types package was split already anyways (`Requirable` vs `Validator`), I'll change it.